### PR TITLE
updated API to version 2.x

### PR DESCRIPTION
**Pull request by mattjackson220**
## What does this PR (Pull Request) do?

This PR updates the API to version 2.0 and promotes all of the most recent API routes to that version.  Backwards compatibility is supported for versions 1.x but requests to 2.x will NOT pass through to 1.x or to the Perl TO API.  It also updates the API to use a struct instead of a float for the version number

- [x] This PR is not related to any Issue 

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

verify that GETs, POSTs, PUTs, and DELETEs to various endpoints work the same when sent to /api/1.5/ and /api/2.0/ .

## The following criteria are ALL met by this PR

Documentation upgrades and further test upgrades are in separate PRs

- [x] This PR includes tests
- [x] I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Comment by ocket8888** [comment]
Why not 2.0?

**Comment by mattjackson220** [comment]
> Why not 2.0?

2.0 doesnt work as is because it changes /api/2.0/ to /api/2/. which could be fixed but with version 1 there wasnt a 1.0 so went this route for consistency

**Comment by ocket8888** [comment]
I don't think you need to allow `/api/2` to support `/api/2.0` - and actually there is an API 1.0, you can see the routes [in TrafficOpsRoutes.pm](https://github.com/apache/trafficcontrol/blob/master/traffic_ops/app/lib/TrafficOpsRoutes.pm#L907-L1005). The thing is that 1.0 was just used to refer to the API before it was versioned, because that was actually our "first" version of the API. So that's why the version-in-the-path API starts at 1.1

**Comment by rawlinp** [comment]
Yeah unfortunately the routing code models the version as a float, so I can see why 2.0 ends up being just 2. Ideally we would refactor that to use the `Version` struct which is just two unsigned ints -- one for major and one for minor -- but I don't _personally_ mind just going to 2.1 to get it out sooner. It's kind of confusing at first, but at the end of the day it's just a number.

**Comment by mhoppa** [comment]
I would vote to spend a little time investigating (maybe time box to a couple of hours) making `2.0` work and if the change seems to be too big or risky I am find as Rawlin said to go with `2.1` to get it out sooner.

**Comment by ocket8888** [comment]
I sort of thought this would be a more involved change such that switching to use the Version structure wouldn't be much additional work. I'd also like to get that to work, since it's something we've wanted for a while now.

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Updated Traffic Router to read new EDSN0 client subnet field and route accordingly only for enabled delivery services. When enabled and a subnet is present in the request, the subnet appears in the `chi` field and the resolver address is in the `rhi` field.
 - Added an optimistic quorum feature to Traffic Monitor to prevent false negative states from propagating to downstream components in the event of network isolation.
 - Traffic Ops Golang Endpoints
+  - /api/2.1 for all of the most recent route versions
```
Comment (mhoppa) [inline CHANGELOG.md]: 2.0

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 2]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+		{api.Version{2, 0}, http.MethodPut, `cachegroups/{id}$`, api.UpdateHandler(&cachegroup.TOCacheGroup{}), auth.PrivLevelOperations, Authenticated, nil, 212954546, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `cachegroups/?$`, api.CreateHandler(&cachegroup.TOCacheGroup{}), auth.PrivLevelOperations, Authenticated, nil, 22982665, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `cachegroups/{id}$`, api.DeleteHandler(&cachegroup.TOCacheGroup{}), auth.PrivLevelOperations, Authenticated, nil, 227869365, noPerlBypass},
+
+		{api.Version{2, 0}, http.MethodPost, `cachegroups/{id}/queue_update$`, cachegroup.QueueUpdates, auth.PrivLevelOperations, Authenticated, nil, 2071644110, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `cachegroups/{id}/deliveryservices/?$`, cachegroup.DSPostHandler, auth.PrivLevelOperations, Authenticated, nil, 2520240431, noPerlBypass},
+
+		//CacheGroup Parameters: CRUD
+		{api.Version{2, 0}, http.MethodGet, `cachegroups/{id}/parameters/?(\.json)?$`, api.ReadHandler(&cachegroupparameter.TOCacheGroupParameter{}), auth.PrivLevelReadOnly, Authenticated, nil, 212449723, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: on merge to master make sure to include cachegroupparameters POST/GET here

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 3]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+		{api.Version{2, 0}, http.MethodGet, `cachegroups/{id}/unassigned_parameters/?(\.json)?$`, api.ReadHandler(&cachegroupparameter.TOCacheGroupUnassignedParameter{}), auth.PrivLevelReadOnly, Authenticated, nil, 2457339250, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `cachegroupparameters/{cachegroupID}/{parameterId}$`, api.DeleteHandler(&cachegroupparameter.TOCacheGroupParameter{}), auth.PrivLevelOperations, Authenticated, nil, 212449733, noPerlBypass},
+
+		//Capabilities
+		{api.Version{2, 0}, http.MethodGet, `capabilities(/|\.json)?$`, capabilities.Read, auth.PrivLevelReadOnly, Authenticated, nil, 2008135, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `capabilities(/|\.json)?$`, capabilities.Create, auth.PrivLevelOperations, Authenticated, nil, 2, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: this should be deprecated remove

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 4]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+
+		//Coordinates
+		{api.Version{2, 0}, http.MethodGet, `coordinates/?(\.json)?$`, api.ReadHandler(&coordinate.TOCoordinate{}), auth.PrivLevelReadOnly, Authenticated, nil, 296700745, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `coordinates/?$`, api.ReadHandler(&coordinate.TOCoordinate{}), auth.PrivLevelReadOnly, Authenticated, nil, 224546706, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPut, `coordinates/?$`, api.UpdateHandler(&coordinate.TOCoordinate{}), auth.PrivLevelOperations, Authenticated, nil, 268926174, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `coordinates/?$`, api.CreateHandler(&coordinate.TOCoordinate{}), auth.PrivLevelOperations, Authenticated, nil, 2428112157, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `coordinates/?$`, api.DeleteHandler(&coordinate.TOCoordinate{}), auth.PrivLevelOperations, Authenticated, nil, 2303849889, noPerlBypass},
+
+		//ASNs
+		{api.Version{2, 0}, http.MethodPut, `asns/?$`, api.UpdateHandler(&asn.TOASNV11{}), auth.PrivLevelOperations, Authenticated, nil, 2264172317, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: nit but should these move up with the other ASNs endpoints

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 5]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+
+		////DeliveryServices
+		{api.Version{2, 0}, http.MethodGet, `deliveryservices/?(\.json)?$`, api.ReadHandler(&deliveryservice.TODeliveryService{}), auth.PrivLevelReadOnly, Authenticated, nil, 2238317294, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `deliveryservices/{id}/?(\.json)?$`, api.ReadHandler(&deliveryservice.TODeliveryService{}), auth.PrivLevelReadOnly, Authenticated, nil, 244348195, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `deliveryservices/?(\.json)?$`, deliveryservice.CreateV15, auth.PrivLevelOperations, Authenticated, nil, 206431432, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPut, `deliveryservices/{id}/?(\.json)?$`, deliveryservice.UpdateV15, auth.PrivLevelOperations, Authenticated, nil, 2766567527, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `deliveryservices/{id}/?(\.json)?$`, api.DeleteHandler(&deliveryservice.TODeliveryService{}), auth.PrivLevelOperations, Authenticated, nil, 222642074, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `deliveryservices/{id}/servers/eligible/?(\.json)?$`, deliveryservice.GetServersEligible, auth.PrivLevelReadOnly, Authenticated, nil, 274761584, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: when merging in master make sure to put deliveryservices/{id}/safe here

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 6]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+		{api.Version{2, 0}, http.MethodPut, `divisions/{id}$`, api.UpdateHandler(&division.TODivision{}), auth.PrivLevelOperations, Authenticated, nil, 206369140, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `divisions/?$`, api.CreateHandler(&division.TODivision{}), auth.PrivLevelOperations, Authenticated, nil, 253713800, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `divisions/{id}$`, api.DeleteHandler(&division.TODivision{}), auth.PrivLevelOperations, Authenticated, nil, 2325382237, noPerlBypass},
+
+		{api.Version{2, 0}, http.MethodGet, `logs/?(\.json)?$`, logs.Get, auth.PrivLevelReadOnly, Authenticated, nil, 248340550, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `logs/{days}/days/?(\.json)?$`, logs.Get, auth.PrivLevelReadOnly, Authenticated, nil, 2192414145, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `logs/newcount/?(\.json)?$`, logs.GetNewCount, auth.PrivLevelReadOnly, Authenticated, nil, 2405833012, noPerlBypass},
+
+		//HWInfo
+		{api.Version{2, 0}, http.MethodGet, `hwinfo/?(\.json)?$`, hwinfo.Get, auth.PrivLevelReadOnly, Authenticated, nil, 221685998, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: this is deprecated remove

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 7]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+
+		// Federation Resolvers
+		{api.Version{2, 0}, http.MethodPost, `federation_resolvers(/|\.json)?$`, federation_resolvers.Create, auth.PrivLevelAdmin, Authenticated, nil, 2134373661, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `federation_resolvers(/|\.json)?$`, federation_resolvers.Read, auth.PrivLevelReadOnly, Authenticated, nil, 256608759, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `federations/{id}/federation_resolvers(/|\.json)?$`, federations.AssignFederationResolversToFederationHandler, auth.PrivLevelAdmin, Authenticated, nil, 256608760, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `federations/{id}/federation_resolvers(/|\.json)?$`, federations.GetFederationFederationResolversHandler, auth.PrivLevelReadOnly, Authenticated, nil, 256608761, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `federation_resolvers(/|\.json)?$`, federation_resolvers.Delete, auth.PrivLevelAdmin, Authenticated, nil, 2001, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `federation_resolvers/{id}(/|\.json)?$`, federation_resolvers.DeleteByID, auth.PrivLevelAdmin, Authenticated, nil, 22, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: this should be deprecated remove

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 8]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+
+		//HWInfo
+		{api.Version{2, 0}, http.MethodGet, `hwinfo/?(\.json)?$`, hwinfo.Get, auth.PrivLevelReadOnly, Authenticated, nil, 221685998, noPerlBypass},
+
+		//Content invalidation jobs
+		{api.Version{2, 0}, http.MethodGet, `jobs(/|\.json/?)?$`, api.ReadHandler(&invalidationjobs.InvalidationJob{}), auth.PrivLevelReadOnly, Authenticated, nil, 2966782041, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `jobs/?$`, invalidationjobs.Delete, auth.PrivLevelPortal, Authenticated, nil, 216780776, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPut, `jobs/?$`, invalidationjobs.Update, auth.PrivLevelPortal, Authenticated, nil, 286134226, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `jobs/?`, invalidationjobs.Create, auth.PrivLevelPortal, Authenticated, nil, 20450955, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `user/current/jobs(/|\.json/?)?$`, invalidationjobs.CreateUserJob, auth.PrivLevelPortal, Authenticated, nil, 211328688, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: deprecated please remove

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 9]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+		{api.Version{2, 0}, http.MethodGet, `user/current/jobs(/|\.json/?)?$`, invalidationjobs.GetUserJobs, auth.PrivLevelReadOnly, Authenticated, nil, 249163540, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: deprecated please remove

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 10]
```diff
@@ -614,7 +1056,7 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
 	// verify configured perl_routes are actually able to pass through to Perl
 	perlRoutes := GetRouteIDMap(d.PerlRoutes)
 	for _, r := range routes {
-		if _, isPerlRoute := perlRoutes[r.ID]; isPerlRoute && !r.CanBypassToPerl {
+		if _, isPerlRoute := perlRoutes[r.ID]; isPerlRoute && (!r.CanBypassToPerl || r.Version.Major > 2) { //do not allow any routes from version 2.x and above to pass thorugh to Perl
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: should the major conditional be ```>=```

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 11]
```diff
@@ -729,7 +1178,18 @@ type ThrottledHandler struct {
...
 func (m ThrottledHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	m.ReqChan <- struct{}{}
+	pathParts := strings.Split(r.URL.Path, "/")
+	if len(pathParts) < 3 {
+		notFoundHandler(w, r)
+		return
+	}
+
+	version, err := strconv.ParseFloat(pathParts[2], 64)
+	if err != nil || version > 2 { // do not default to Perl for versions over 2.x
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: should version be  version be ```>= 2```

File: traffic_ops/traffic_ops_golang/routing/routing.go [hunk 12]
```diff
@@ -59,7 +60,7 @@ type Route struct {
 }
 
 func (r Route) String() string {
-	return fmt.Sprintf("id=%d method=%s version=%.1f path=%s can_bypass_to_perl=%t", r.ID, r.Method, r.Version, r.Path, r.CanBypassToPerl)
+	return fmt.Sprintf("id=%d method=%s version=%d.%d.1f path=%s can_bypass_to_perl=%t", r.ID, r.Method, r.Version.Major, r.Version.Minor, r.Path, r.CanBypassToPerl)
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routing.go]: see rawlin's comment should not need floating point precision anymore

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 13]
```diff
@@ -133,472 +134,913 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+		{api.Version{2, 0}, http.MethodDelete, `tenants/{id}$`, api.DeleteHandler(&apitenant.TOTenant{}), auth.PrivLevelOperations, Authenticated, nil, 216365558, noPerlBypass},
+
+		//CRConfig
+		{api.Version{2, 0}, http.MethodGet, `cdns/{cdn}/snapshot/?$`, crconfig.SnapshotGetHandler, auth.PrivLevelReadOnly, Authenticated, nil, 2957273695, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `cdns/{cdn}/snapshot/new/?$`, crconfig.Handler, auth.PrivLevelReadOnly, Authenticated, nil, 276716889, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPut, `cdns/{id}/snapshot/?$`, crconfig.SnapshotHandler, auth.PrivLevelOperations, Authenticated, nil, 254424150, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPut, `snapshot/{cdn}/?$`, crconfig.SnapshotHandler, auth.PrivLevelOperations, Authenticated, nil, 2969911829, noPerlBypass},
+
+		// ATS config files
+		{api.Version{2, 0}, http.MethodGet, `servers/{server-name-or-id}/configfiles/ats/?(\.json)?$`, atsserver.GetConfigMetaData, auth.PrivLevelOperations, Authenticated, nil, 2755842214, noPerlBypass},
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: I don't think we want to promote the ATS configfile endpoints to 2.0. There was a mailing list discussion about it a while back, and `atstccfg` is going to be responsible for ATS config gen going forward. Is that right, @rob05c?

Comment (rob05c) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: Yes, if we're rolling the major version, we don't want to keep supporting the config endpoints, they're no longer used by mainline TO.

There's been some interest in continuing to generate config in TO on the mailing list. But ORT doesn't use them anymore. I think if someone wants to have central/TO config generation, they can make a TO Plugin calling `lib/go-atstccfg` (and I'm +1 on merging that, if we have users willing to write and maintain it. Plugins have a lower maintenance and support bar, and don't fall under the API Versioning promise [unless the Plugin maintainer wants to make a promise]).

File: CHANGELOG.md (hunk 1 above)
Comment (mattjackson220) [reply CHANGELOG.md]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 3 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 4 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 6 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 7 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 8 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 9 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 10 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: yup done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 11 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routing.go (hunk 12 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routing.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go (hunk 13 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 14]
```diff
@@ -133,477 +134,872 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+		{api.Version{2, 0}, http.MethodPost, `roles/?$`, api.CreateHandler(&role.TORole{}), auth.PrivLevelAdmin, Authenticated, nil, 230652406, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `roles/?$`, api.DeleteHandler(&role.TORole{}), auth.PrivLevelAdmin, Authenticated, nil, 2356705982, noPerlBypass},
+
+		//Delivery Services Regexes
+		{api.Version{2, 0}, http.MethodGet, `deliveryservices_regexes/?$`, deliveryservicesregexes.Get, auth.PrivLevelReadOnly, Authenticated, nil, 205501453, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `deliveryservices/{dsid}/regexes/?$`, deliveryservicesregexes.DSGet, auth.PrivLevelReadOnly, Authenticated, nil, 277432763, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `deliveryservices/{dsid}/regexes/{regexid}?$`, deliveryservicesregexes.DSGetID, auth.PrivLevelReadOnly, Authenticated, nil, 2044974567, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `deliveryservices/{dsid}/regexes/?(\.json)?$`, deliveryservicesregexes.Post, auth.PrivLevelOperations, Authenticated, nil, 212737800, noPerlBypass},
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 15]
```diff
@@ -133,477 +134,872 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+		{api.Version{2, 0}, http.MethodGet, `profiles/{id}/unassigned_parameters/?$`, profileparameter.GetUnassigned, auth.PrivLevelReadOnly, Authenticated, nil, 274429262, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `profiles/name/{name}/parameters/?$`, profileparameter.GetProfileName, auth.PrivLevelReadOnly, Authenticated, nil, 2267737832, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `parameters/profile/{name}/?$`, profileparameter.GetProfileName, auth.PrivLevelReadOnly, Authenticated, nil, 2802599194, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `profiles/name/{name}/parameters/?$`, profileparameter.PostProfileParamsByName, auth.PrivLevelOperations, Authenticated, nil, 2355945582, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `profiles/{id}/parameters/?$`, profileparameter.PostProfileParamsByID, auth.PrivLevelOperations, Authenticated, nil, 216818708, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `profileparameters/?(\.json)?$`, api.ReadHandler(&profileparameter.TOProfileParameter{}), auth.PrivLevelReadOnly, Authenticated, nil, 250609805, noPerlBypass},
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 16]
```diff
@@ -133,477 +134,872 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
...
+
+		//Region: CRUDs
+		{api.Version{2, 0}, http.MethodGet, `regions/?$`, api.ReadHandler(&region.TORegion{}), auth.PrivLevelReadOnly, Authenticated, nil, 210037085, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `regions/{id}$`, api.ReadHandler(&region.TORegion{}), auth.PrivLevelReadOnly, Authenticated, nil, 2224440051, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `regions/name/{name}/?$`, region.GetName, auth.PrivLevelReadOnly, Authenticated, nil, 203583197, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPut, `regions/{id}$`, api.UpdateHandler(&region.TORegion{}), auth.PrivLevelOperations, Authenticated, nil, 222308224, noPerlBypass},
+		{api.Version{2, 0}, http.MethodPost, `regions/?$`, api.CreateHandler(&region.TORegion{}), auth.PrivLevelOperations, Authenticated, nil, 2288334488, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `regions/?$`, api.DeleteHandler(&region.TORegion{}), auth.PrivLevelOperations, Authenticated, nil, 2232626758, noPerlBypass},
+		{api.Version{2, 0}, http.MethodDelete, `regions/name/{name}$`, handlerToFunc(proxyHandler), 0, NoAuth, nil, 2925881096, noPerlBypass},
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 17]
```diff
@@ -722,6 +1118,13 @@ func notImplementedHandler(w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(http.StatusText(code)))
 }
 
+// notFoundHandler returns a 404 Not Found to the client.  This should be the new rootHandler for API version 2.x in order to prevent falling back to Perl.
+func notFoundHandler(w http.ResponseWriter, r *http.Request) {
+	code := http.StatusNotFound
+	w.WriteHeader(code)
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/routing/routes.go]: done

**Pull request merged**